pkg/cache/multi_namespace_cache.go

```diff
@@ -37,37 +39,153 @@ type NewCacheFunc func(config *rest.Config, opts Options) (Cache, error)
+		options.NewDefaultNamespacedCache = f
+	}
+}
+
+// MultiNamespacedCacheWithOptionsBuilder builds a composite cache that delegates to per-namespace
+// caches built according to the passed MultiNamespacedOption options.
+//
+// If the set of options passed to MultiNamespacedCacheWithOptionsBuilder results in the NewCacheFunc
+// being set multiple times for the same namespace, the last such setting will be used.
+//
+// If a default namespaced cache is defined (e.g. with WithDefaultNamespacedCache), it will be used
+// as a catch-all for objects not in the explicit namespaces configured on the cache builder. The
+// default namespaced cache is automatically configured with extra field selectors to avoid duplicate
+// caching of objects between namespace-specific caches and this catch-all cache.
+//
+// If a cluster-scoped cache is defined (e.g. with WithClusterScopedCache), it will be used for
+// cluster-scoped objects. If it is undefined, the resulting cache will return an error during read
+// operations, reporting that a cluster-scoped cache is not defined.
+func MultiNamespacedCacheWithOptionsBuilder(opts ...MultiNamespacedOption) NewCacheFunc {
+	multiNamespaceOpts := MultiNamespacedOptions{
+		NewNamespaceCaches: map[string]NewCacheFunc{},
+	}
+	for _, opt := range opts {
+		opt(&multiNamespaceOpts)
+	}
+
 	return func(config *rest.Config, opts Options) (Cache, error) {
 		opts, err := defaultOpts(config, opts)
 		if err != nil {
 			return nil, err
 		}
 
-		caches := map[string]Cache{}
+		var clusterCache Cache
+		if multiNamespaceOpts.NewClusterScopedCache != nil {
+			clusterCache, err = multiNamespaceOpts.NewClusterScopedCache(config, opts)
+			if err != nil {
+				return nil, err
+			}
+		}
 
-		// create a cache for cluster scoped resources
-		gCache, err := New(config, opts)
-		if err != nil {
-			return nil, fmt.Errorf("error creating global cache: %w", err)
+		nsToCache := map[string]Cache{}
+		if multiNamespaceOpts.NewDefaultNamespacedCache != nil {
+			defaultNamespaceCache, err := multiNamespaceOpts.NewDefaultNamespacedCache(config, ignoreNamespaces(opts, multiNamespaceOpts.NewNamespaceCaches))
+			if err != nil {
+				return nil, err
+			}
+			nsToCache[corev1.NamespaceAll] = defaultNamespaceCache
 		}
 
-		for _, ns := range namespaces {
+		for ns, newCacheFunc := range multiNamespaceOpts.NewNamespaceCaches {
 			opts.Namespace = ns
-			c, err := New(config, opts)
+			nsToCache[ns], err = newCacheFunc(config, opts)
 			if err != nil {
 				return nil, err
 			}
-			caches[ns] = c
 		}
-		return &multiNamespaceCache{namespaceToCache: caches, Scheme: opts.Scheme, RESTMapper: opts.Mapper, clusterCache: gCache}, nil
+
+		return &multiNamespaceCache{
+			namespaceToCache: nsToCache,
+			clusterCache:     clusterCache,
+			RESTMapper:       opts.Mapper,
+			Scheme:           opts.Scheme,
+		}, nil
+	}
+}
+
+func ignoreNamespaces(opts Options, newObjectCaches map[string]NewCacheFunc) Options {
```

Comment:
Not sure if the func name matches the implementation given that we also add opts.DefaultSelector.Field

Maybe something a bit more generic like `calculateDefaultNamespacedCacheOpts` (just an idea)

(Assuming I interpret the code correctly and the selector is not only applied on namespaces but also on other object types)